# Project 3
### By Ishaan Nagpal and Cameron Schofield
### Working hours: 9 hours
## Exercise 1

In [1]:
# Hamming distance between x and y
def hamming(x, y):
    d = 0

    # Count number of mismatches
    for x_i, y_i in zip(x, y):
        if x_i != y_i:
            d += 1

    return d    


# LFSR with coefficients c and initial state s
def LFSR(c, s, L, N):
    # Generate first N states
    while len(s) != N:
        # Calculate next state
        next_state = 0
        for i in range(L):
            next_state -= c[i] * s[len(s)-i-1]
        
        next_state %= 2

        s += [next_state]

    return s

# G
def generate(z, c):
    L = len(c)
    N = len(z)
    
    # Store minimum probability and state that achievee it
    min_p = 1
    min_s = []

    # Test all possible starting states
    for k in range(pow(2, L)):
        # Convert integer k to bit string
        b = format(k, "b").zfill(L)
        # Create initial state array
        s = []
        for char in b:
            if char == "0":
                s.append(0)
            else:
                s.append(1)
        
        # Compute first N states using LFSR
        u = LFSR(c, s, L, N)

        # Compute probability
        p = hamming(u, z) / N

        # Find minimum probability
        if p <= min_p:
            min_p = p
            min_s = s
    
    # Print probability of minimizing state
    print(1 - min_p)
    return min_s


if __name__ == '__main__':
    # Keystream, convert to array
    z_txt = "1110000100101110110001101001000110010000010100110101100101101011110100010001001010110010111111011011000100011011001101011110110111110000111001101011001111110010110000111011010011011111111011001"
    z = []
    for char in z_txt:
        z.append(int(char))
    
    # Coefficients for each LFSR
    c1 = [1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1]
    c2 = [0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1]
    c3 = [0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1]

    # Generate first N (length of keystream) states from each LFSR
    t1 = generate(z, c1)
    t2 = generate(z, c2)
    t3 = generate(z, c3)

    # Print starting states for each LFSR
    print(t1[:13])
    print(t2[:15])
    print(t3[:17])

    # Check for errors
    for i1, i2, i3, x in zip(t1, t2, t3, z):
        if i1 + i2 + i3 >= 2:
            if x == 0:
                print("wrong")
        else:
            if x == 1:
                print("wrong")


0.766839378238342
0.7305699481865284
0.7616580310880829
[1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1]
[1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0]
[1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0]


#### Initial states of the 3 LFSRs
$K = \left(1110000100111, 101011010110000, 11100000000011100 \right)$

#### Maximized probabilities for initial states
1: 0.766839378238342  
2: 0.7305699481865284  
3: 0.7616580310880829  


## Exercise 2
It takes T seconds to check $2^{13}+2^{15}+2^{17}$ initial states in our attack. It takes $2^{45}$ checks to do exhaustive search over the entire keyspace. This means it takes $\frac{2^{45}}{2^{13}+2^{15}+2^{17}}*T$ seconds for exhaustive search.